src/py/mat3ra/made/tools/build/defect/builders.py

```diff
@@ -173,23 +209,102 @@ def create_adatom(
     def get_equidistant_position(
         self, material: Material, position_on_surface: List[float], distance_z: float = 2.0
     ) -> List[float]:
-        new_basis = material.basis
-        adatom_position = self._calculate_position_from_2d(material, position_on_surface, distance_z)
-        neighboring_atoms_ids = get_nearest_neighbors_atom_indices(material, adatom_position)
-        # We need to check if neighboring atoms number is the same in pbc
+        adatom_coordinate = self._calculate_coordinate_from_position_and_distance(
+            material, position_on_surface, distance_z
+        )
+
+        # We need to find the neighboring atoms with pbc.
         supercell_material = create_supercell(material, [[3, 0, 0], [0, 3, 0], [0, 0, 1]])
         # Move the coordinate to the central unit cell of the supercell (crystal coordinates)
-        supercell_adatom_position = [1 / 3 + adatom_position[0] / 3, 1 / 3 + adatom_position[1] / 3, adatom_position[2]]
+        supercell_adatom_coordinate = [
+            1 / 3 + adatom_coordinate[0] / 3,
+            1 / 3 + adatom_coordinate[1] / 3,
+            adatom_coordinate[2],
+        ]
         supercell_neighboring_atoms_ids = get_nearest_neighbors_atom_indices(
-            supercell_material, supercell_adatom_position
+            supercell_material, supercell_adatom_coordinate
         )
-        if neighboring_atoms_ids is None or supercell_neighboring_atoms_ids is None:
+
+        if supercell_neighboring_atoms_ids is None:
             raise ValueError("No neighboring atoms found. Try reducing the distance_z.")
-        if len(supercell_neighboring_atoms_ids) != len(neighboring_atoms_ids):
-            raise ValueError("Number of neighboring atoms is not the same in PBC. Try increasing the supercell size.")
-        neighboring_atoms_coordinates = [new_basis.coordinates.values[atom_id] for atom_id in neighboring_atoms_ids]
 
-        equidistant_position = get_center_of_coordinates(neighboring_atoms_coordinates)
-        equidistant_position[2] = adatom_position[2]
+        new_supercell_basis = supercell_material.basis.copy()
+        new_supercell_basis.coordinates.filter_by_ids(supercell_neighboring_atoms_ids)
+        neighboring_atoms_coordinates = new_supercell_basis.coordinates.values
+        supercell_equidistant_coordinate = get_center_of_coordinates(neighboring_atoms_coordinates)
+        supercell_equidistant_coordinate[2] = adatom_coordinate[2]
+
+        equidistant_coordinate = [
```

Comment:
This and the above functions converting to and from the 3x3x1 supercell can go to utils to save some space and assist with readability